### PR TITLE
Improve multi-executors RuleTypes flow

### PR DESCRIPTION
**Pull request by rafaelverger**
This changes makes "how to pause executors" as explicit as possible.

### CODE REVIEW

File: src/core/controller.ts [hunk 1]
```diff
@@ -208,24 +238,19 @@ export class Controller {
...
-      if ( this.getRuleExecutorIndex(rule) < executors.length-1 ) {
-        this.setRuleExecutorIndex(rule, this.getRuleExecutorIndex(rule)+1);
-        bot.dispatch('hear');
-        return this;
-      }
-
-      this.setRuleExecutorIndex(rule);
+      answer = await this.executeRuleTypeExecutors(rule, message);
     } catch (e) {
+      let expectedError = false;
```
Comment (andersonba) [inline src/core/controller.ts]: [minor] renames to `expectedReply`

File: src/core/exceptions.ts [hunk 2]
```diff
@@ -20,3 +20,8 @@ export function RuleNotFound(name: string, indexes: { [k: string]: number }) {
   this.message = `Rule "${name}" not found in indexes`;
   this.indexes = indexes;
 }
+
+export function PauseRuleTypeExecutors(name: string) {
+  this.key = 'pauseRuleTypeExecutors';
+  this.message = `Rule "${name}" should pause exection`;
```
Comment (andersonba) [inline src/core/exceptions.ts]: typo

File: src/types.ts [hunk 3]
```diff
@@ -1,8 +1,19 @@
...
   rule?: Rule;
 }
 
+export interface RuleType {
+  executors: RuleTypeExecutor[];
+}
+
+export interface RuleTypeExecutor {
+  validators?: RuleValidator[];
+  transform?: (value: any, rule?: Rule, bot?: YveBot) => Promise<any>;
```
Comment (andersonba) [inline src/types.ts]: create a type `RuleTransform`

File: src/core/bot.ts [hunk 4]
```diff
@@ -121,3 +126,7 @@ export class YveBot {
 YveBot.prototype.types = new Types;
 YveBot.prototype.actions = new Actions;
 YveBot.prototype.validators = new Validators;
+
+
+YveBot.Exceptions = Exceptions;
```
Comment (andersonba) [inline src/core/bot.ts]: use `exceptions` to keep the pattern
`YveBot.exceptions.RuleNotFound`

File: src/core/bot.ts [hunk 5]
```diff
@@ -121,3 +126,7 @@ export class YveBot {
...
+YveBot.WaitForUserInputExecutor = WaitForUserInput;
```
Comment (andersonba) [inline src/core/bot.ts]: `YveBot.executors = new Executors`
to be accessed by `YveBot.executors.WaitForUserInput`

File: src/core/types.ts [hunk 6]
```diff
@@ -1,9 +1,19 @@
 import { uniq } from 'lodash';
 import { DefineModule } from './module';
-import { Rule, Answer } from '../types';
+import { PauseRuleTypeExecutors } from './exceptions';
+import { Rule, Answer, RuleType, RuleTypeExecutor } from '../types';
 import * as utils from './utils';
 
-const types = {
+
+export const WaitForUserInput: RuleTypeExecutor = {
```
Comment (andersonba) [inline src/core/types.ts]: move it to `./executors`

File: src/core/bot.ts [hunk 7]
```diff
@@ -23,6 +25,9 @@ function sanitizeRule(rule: Rule): Rule {
 }
 
 export class YveBot {
+  static Exceptions: any;
```
Comment (andersonba) [inline src/core/bot.ts]: Exceptions instead any

File: src/core/bot.ts [hunk 8]
```diff
@@ -23,6 +25,9 @@ function sanitizeRule(rule: Rule): Rule {
...
+  static WaitForUserInputExecutor: typeof WaitForUserInput;
```
Comment (andersonba) [inline src/core/bot.ts]: try without typeof

File: src/core/controller.ts [hunk 9]
```diff
@@ -190,12 +196,36 @@ export class Controller {
...
+    this.bot.store.set(
+      `executors.${rule.name}.currentIdx`, this.getRuleExecutorIndex(rule) + 1
+    );
+  }
+
+  resetRuleExecutorIndex(rule: Rule): void {
+    this.bot.store.unset(`executors.${rule.name}.currentIdx`);
+  }
+
+  async executeRuleTypeExecutors(rule: Rule, lastAnswer: Answer | Answer[]): Promise<any> {
```
Comment (andersonba) [inline src/core/controller.ts]: `Promise<Answer | Answer[]>` breaks?

Comment (rafaelverger) [reply src/core/controller.ts]: There is no guarantee that return will be answer... the return will be whatever transform + validators returns

Comment (rafaelverger) [reply src/core/controller.ts]: done

**Pull request merged**